src/App.js

```diff
@@ -1,79 +1,104 @@
+  useEffect(() => {
+    // For testing and debugging purposes
+    console.log("MTURK? :" + MTURK);
+    console.log("ELECTRON? :" + IS_ELECTRON);
+    console.log("AT_HOME? :" + AT_HOME);
+    console.log("FIREBASE? :" + FIREBASE);
+
+    // If on desktop
+    if (IS_ELECTRON) {
+      const electron = window.require("electron");
+      const renderer = electron.ipcRenderer;
+      setRenderer(renderer);
+      // If at home, fill in fields based on environment variables
+      if (AT_HOME) {
+        const credentials = renderer.sendSync("syncCredentials");
+        if (credentials.envParticipantId != null) {
+          setEnvParticipantId(credentials.envParticipantId);
+        }
+        if (credentials.envStudyId != null) {
+          setEnvStudyId(credentials.envStudyId);
+        }
+      }
+      // If in clinic - currently placeholder
+      else {
+
+      }
+      setValidation("desktop");
+      // If online
+    } else {
+      // If MTURK
+      if (MTURK) {
+        /* eslint-disable */
+        window.lodash = _.noConflict();
+        setPsiturk(new PsiTurk(getTurkUniqueId(), "/complete"));
+        setLogin(true);
+        /* eslint-enable */
+      }
+      // If firebase
+      else if (FIREBASE) {
+        // If prolific - currectly placeholder
+        if (PROLIFIC) {
+
+        }
+        // Otherwise - currently placeholder
+        else {
+          
+        }
+        setValidation("firebase");
+      }
+    }
+  }, []);
 
-class App extends React.Component {
-  render() {
-    console.log("Outside Turk:", jsPsych.turk.turkInfo().outsideTurk)
-    console.log("Turk:", MTURK)
-    jsPsych.plugins['rt-categorize-html'] = rt_categorize_html();
-    
-    return (
-      <div className="App">
-        <Experiment settings={{
-          timeline: tl,
-          on_data_update: (data) => {
-            //firebase 
-            if(FIREBASE){
-              if (data.trial_index === 1) {
-                
-                console.log(data.patient_id);// delete when done
-  
-                createFirebaseDocument(data.patient_id);
-                addToFirebase(data);
-              }
-              if (data.trial_index > 1) {
-                addToFirebase(data);
-              }
-            }
-            else if ( ipcRenderer ) {
-              ipcRenderer.send('data', data)
-            }
-            else if (psiturk) {
-                psiturk.recordTrialData(data)
-            }
-          },
-          on_finish: (data) => {
-            if ( ipcRenderer ) {
-              ipcRenderer.send('end', 'true')
-            }
-            else if (psiturk) {
-              const completePsiturk = async () => {
-                psiturk.saveData()
-                await sleep(5000)
-                psiturk.completeHIT()
-              }
-              completePsiturk()
-            }
-          },
-        }}
+  return (
+    <>
+      {loggedIn ? (
+        <JsPsychExperiment ipcRenderer={ipcRenderer} psiturk={psiturk} />
```

Comment:
I think it's cleaner to pass a function instead of the `ipcRenderer` object. The `ipcRenderer` object should be used as little as possible as it allows for low-level access to everything in electron. It's so dangerous that in later versions of Electron they disable it by default.

Perhaps send a function as a prop to `JsPsychExperiment` that either sends data via `ipcRenderer` or records data via `psiturk`. You may need two functions there, one for data update and one for finish

Reply:
I see. This is what was there in the original App.js so I didn't want to change the things that worked. I'll make these changes and make it similar to the validation function and make sure eveyrthing works.